File: src/linear.js
import { LINEAR_API_KEY, LINEAR_TEAM_ID } from './config.js';

const GRAPHQL_ENDPOINT = 'https://api.linear.app/graphql';

const CREATE_ISSUE_MUTATION = `
mutation CreateIssue($input: IssueCreateInput!) {
  issueCreate(input: $input) {
    success
    issue {
      id
      identifier
      title
      url
    }
  }
}
`;

async function gql(query, variables = {}) {
  const resp = await fetch(GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers: {
      'Authorization': LINEAR_API_KEY,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query, variables }),
  });
  if (!resp.ok) throw new Error(`Linear API HTTP ${resp.status}`);
  const data = await resp.json();
  if (data.errors) throw new Error(`Linear API error: ${JSON.stringify(data.errors)}`);
  return data;
}

export async function getTeams() {
  if (!LINEAR_API_KEY) return [];
  try {
    const data = await gql('{ teams { nodes { id name key } } }');
    return data.data.teams.nodes;
  } catch {
    return [];
  }
}

export async function getTeamName(teamId) {
  const teams = await getTeams();
  const team = teams.find(t => t.id === teamId);
  return team ? `${team.name} (${team.key})` : null;
}

export async function createIssue(ticket, teamId = null) {
  if (!LINEAR_API_KEY) throw new Error('LINEAR_API_KEY is not set. Check your .env file.');
  const team = teamId || LINEAR_TEAM_ID;
  if (!team) throw new Error('LINEAR_TEAM_ID is not set. Check your .env file.');

  const data = await gql(CREATE_ISSUE_MUTATION, {
    input: {
      teamId: team,
      title: ticket.title,
      description: ticket.description || '',
      priority: ticket.priority || 3,
    },
  });

  const issue = data.data.issueCreate.issue;
  return {
    id: issue.identifier,
    title: issue.title,
    url: issue.url,
  };
}

File: relay/linear.py
import requests
from relay.config import LINEAR_API_KEY, LINEAR_TEAM_ID

GRAPHQL_ENDPOINT = "https://api.linear.app/graphql"

CREATE_ISSUE_MUTATION = """
mutation CreateIssue($input: IssueCreateInput!) {
  issueCreate(input: $input) {
    success
    issue {
      id
      identifier
      title
      url
    }
  }
}
"""


def get_teams() -> list[dict]:
    """Fetch all teams the user has access to."""
    if not LINEAR_API_KEY:
        return []
    resp = requests.post(
        GRAPHQL_ENDPOINT,
        json={"query": "{ teams { nodes { id name key } } }"},
        headers={"Authorization": LINEAR_API_KEY, "Content-Type": "application/json"},
        timeout=15,
    )
    if resp.status_code != 200:
        return []
    data = resp.json()
    return data.get("data", {}).get("teams", {}).get("nodes", [])


def get_team_name(team_id: str) -> str | None:
    """Get team name by ID."""
    for t in get_teams():
        if t["id"] == team_id:
            return f"{t['name']} ({t['key']})"
    return None


def create_issue(ticket: dict, team_id: str = None) -> dict:
    if not LINEAR_API_KEY:
        raise RuntimeError("LINEAR_API_KEY is not set. Check your .env file.")
    team = team_id or LINEAR_TEAM_ID
    if not team:
        raise RuntimeError("LINEAR_TEAM_ID is not set. Check your .env file.")

    variables = {
        "input": {
            "teamId": team,
            "title": ticket["title"],
            "description": ticket.get("description", ""),
            "priority": ticket.get("priority", 3),
        }
    }

    resp = requests.post(
        GRAPHQL_ENDPOINT,
        json={"query": CREATE_ISSUE_MUTATION, "variables": variables},
        headers={
            "Authorization": LINEAR_API_KEY,
            "Content-Type": "application/json",
        },
        timeout=30,
    )
    resp.raise_for_status()
    data = resp.json()

    if "errors" in data:
        raise RuntimeError(f"Linear API error: {data['errors']}")

    issue = data["data"]["issueCreate"]["issue"]
    return {
        "id": issue["identifier"],
        "title": issue["title"],
        "url": issue["url"],
    }
